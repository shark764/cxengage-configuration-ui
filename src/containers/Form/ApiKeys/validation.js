/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import { isEmpty } from 'serenova-js-utils/strings';

export const formValidation = values => ({
  name: isEmpty(values.get('name')) && 'Please enter a name',
  roleId: !values.get('roleId') && 'You must select a role'
});
