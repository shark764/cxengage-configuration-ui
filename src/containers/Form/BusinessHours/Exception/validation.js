export const formValidation = values => ({
  startTimeMinutes:
    values.get('startTimeMinutes') > values.get('endTimeMinutes') &&
    values.get('endTimeMinutes') !== 0 &&
    'Start time should be ahead of end time.',
  date:
    values.get('date') &&
    new Date(
      values
        .get('date')
        .toJSON()
        .split('T')[0]
    ) < new Date(new Date().toJSON().split('T')[0]) &&
    'Please select a future date'
});
