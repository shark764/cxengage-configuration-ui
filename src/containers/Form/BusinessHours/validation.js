/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import { isEmpty } from 'serenova-js-utils/strings';

export const formValidation = values => {
  const daysInitials = ['sun', 'sat', 'mon', 'tue', 'thu', 'wed', 'fri'];
  let validation = {};

  validation.name = isEmpty(values.get('name')) && 'Please enter a name';
  return daysInitials.reduce(
    (r, i) => ({
      ...r,
      [i.concat('StartTimeMinutes')]:
        values.get(i.concat('StartTimeMinutes')) > values.get(i.concat('EndTimeMinutes')) &&
        values.get(i.concat('EndTimeMinutes')) !== 0 &&
        'Start time may not be after end time'
    }),
    validation
  );
};
