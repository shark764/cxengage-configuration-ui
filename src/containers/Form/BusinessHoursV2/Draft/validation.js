import { isEmpty } from 'serenova-js-utils/strings';

export const formValidation = values => ({
  name: isEmpty(values.get('name')) && 'Please enter a name'
});