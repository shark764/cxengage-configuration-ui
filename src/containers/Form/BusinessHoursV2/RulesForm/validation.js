import { isEmpty } from 'serenova-js-utils/strings';

const repeats = {
  weekly: 'every how many weeks',
  daily: 'every how many days',
  monthly: 'every how many months'
};

const daysOnMonth = {
  january: 31,
  febraury: 29, //Max number fo days febraury can have (leap year)
  march: 31,
  april: 30,
  may: 31,
  june: 30,
  july: 31,
  august: 31,
  september: 30,
  october: 31,
  november: 30,
  december: 31
};

export const formValidation = values => {
  let errors = {};
  let rulesErrors = [];
  if (values.get('rules') && values.get('rules').size) {
    values.get('rules').forEach((rule, index, rules) => {
      const error = ruleValidation(rule, rules);
      if (error) {
        rulesErrors[index] = error;
      }
    });
  }

  if (rulesErrors.length) {
    errors.rules = rulesErrors;
  }
  return errors;
};

export const ruleValidation = (rule, rules) => {
  let error;
  if (isEmpty(rule.name)) {
    error = {
      name: true,
      message: 'The rule should have a name'
    };
  } else if (rules.some(r => r.name.trim() === rule.name.trim() && !isEmpty(rule.name) && r.id !== rule.id)) {
    error = {
      name: true,
      message: 'Rule name should be unique within the draft'
    };
  } else if (isEmpty(rule.type)) {
    error = {
      type: true,
      message: 'The rule should have a type'
    };
  } else if (!rule.startDate) {
    error = {
      startDate: true,
      message: 'You should select a start date'
    };
  } else if (rule.startDate && !rule.type.includes('one-time') && rule.endDate && rule.startDate > rule.endtDate) {
    error = {
      endDate: true,
      message: 'end date should be after start date'
    };
  } else if (!rule.type.includes('one-time') && !rule.repeats) {
    error = {
      repeats: true,
      message: 'You should select the rate at which the rule repeats'
    };
  } else if (!rule.type.includes('one-time') && !rule.every) {
    error = {
      every: true,
      message: `You should set ${repeats[rule.repeats]} should the rule take effect`
    };
  } else if (!rule.type.includes('one-time') && isNaN(rule.every) && rule.repeats !== 'yearly') {
    error = {
      every: true,
      message: 'The Every field should be a number'
    };
  } else if (!rule.type.includes('one-time') && !Number.isInteger(+rule.every) && rule.repeats !== 'yearly') {
    error = {
      every: true,
      message: 'The Every field should be a integer value'
    };
  } 
  else if (!rule.type.includes('one-time') && rule.repeats === 'weekly' && (!rule.on || !Array.isArray(rule.on) || rule.on.length < 1)) {
    error = {
      on: true,
      message: 'You should select the days on the week in which the rule should take effect'
    };
  } else if (!rule.type.includes('one-time') &&
    ((rule.repeats === 'monthly' || rule.repeats === 'yearly') && (!rule.on || !rule.on.type || !rule.on.value)))  {
    error = {
      on: {
        type: !rule.on || !rule.on.type,
        value: !rule.on || !rule.on.value
      },
      message: 'type or value on the "On" field is undefined'
    };
  } else if (!rule.type.includes('one-time') &&
  ((rule.repeats === 'monthly' || rule.repeats === 'yearly') && 
  (rule.on &&
    rule.on.value === 'day' &&
    (rule.on && (isNaN(rule.on.type) || !Number.isInteger(+rule.on.type)))))) {
      error = {
        on: {
          type: true
        },
        message: 'value on the "On" field should be a number'
      };
  } else if (
    !rule.type.includes('one-time') &&
    rule.repeats === 'monthly' &&
    rule.on &&
    rule.on.value === 'day' &&
    (rule.on.type < 1 || rule.on.type > 31)
  ) {
    error = {
      on: {
        type: true,
      },
      message: 'The numeric value in the "On" field has to be between 1 and 31(days on a month)'
    };
  } else if (
    !rule.type.includes('one-time') &&
    rule.repeats === 'yearly' &&
    rule.on &&
    rule.on.value === 'day' &&
    rule.every &&
    (rule.on.type < 1 || rule.on.type > daysOnMonth[rule.every])
  ) {
    error = {
      on: {
        type: true
      },
      errorMessage: `The numeric value in the "On" field has to be between 1 and ${
        daysOnMonth[rule.every]
      } (Number of days on ${rule.every}, considering a day more when leap year on Febraury)`
    };
  } else if (
    rule.hours &&
    !rule.hours.allDay &&
    rule.hours.intervals &&
    rule.hours.intervals.findIndex(
      (interval, ind, intervals) =>
        interval.start > interval.end ||
        intervals.some(
          (int, i) =>
            i !== ind &&
            ((interval.start >= int.start && interval.start <= int.end) ||
              (interval.end >= int.start && interval.end <= int.end))
        )
    ) > -1
  ) {
    const index = rule.hours.intervals.findIndex(
      (interval, ind, intervals) =>
        interval.start > interval.end ||
        intervals.some((int, i) => i !== ind && ((interval.start >= int.start && interval.start <= int.end) || 
          (interval.end >= int.start && interval.end <= int.end)))
    );
    let intervalErrors = Array(rule.hours.intervals.length).fill({
      start: false,
      end: false
    });
    intervalErrors[index] = {
      start: true,
      end: true
    };

    error = {
      hours: {
        intervals: intervalErrors
      },
      message: 'An interval has a greater start time than its end time'
    };
  }

  return error;
}
