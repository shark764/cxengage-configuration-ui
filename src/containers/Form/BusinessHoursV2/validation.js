/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import { isEmpty } from 'serenova-js-utils/strings';

import store from '../../../redux/store';
import { getAllEntitiesTableData } from '../../../containers/EntityTable/selectors';
import { getCurrentEntity } from '../../../redux/modules/entities/selectors';
import { getCurrentTenantId } from '../../../redux/modules/userData/selectors';

export const formValidation = values => ({
  name:
    (isEmpty(values.get('name')) && 'Please enter a name') ||
    (getCurrentEntity(store.getState()) === 'businessHoursV2' &&
      getAllEntitiesTableData(store.getState())
        .filter(({ tenantId }) => tenantId === getCurrentTenantId(store.getState()))
        .some(({ name, id }) => name.toLowerCase() === values.get('name').toLowerCase() && id !== values.get('id')) &&
      "There's already a business hour with this name")
});
