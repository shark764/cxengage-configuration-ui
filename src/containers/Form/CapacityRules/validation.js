/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import { isEmpty } from 'serenova-js-utils/strings';

import store from '../../../redux/store';
import { getEntityData } from '../../../redux/modules/entities/selectors';

export const formValidation = (values, { intl: { formatMessage } }) => ({
  name:
    (isEmpty(values.get('name')) &&
      formatMessage({
        id: 'identityProviders.details.name.error',
        defaultMessage: 'Please enter a name...',
      })) ||
    (getEntityData(store.getState(), 'capacityRules').some(
      (capacityRule) => capacityRule.get('id') !== values.get('id') && capacityRule.get('name') === values.get('name')
    ) &&
      formatMessage({
        id: 'capacityRules.details.name.duplicateError',
        defaultMessage: "Two Capacity Rules can't have the same name",
      })),
});
