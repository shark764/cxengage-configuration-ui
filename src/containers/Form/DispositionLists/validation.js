/*
 * Copyright Â© 2015-2019 Serenova, LLC. All rights reserved.
 */

import { isEmpty } from 'serenova-js-utils/strings';

export const formValidation = (values, props = {}) => {
  let validation = {};
  let { isSaving, isFetching, initialized } = props;
  validation.name = isEmpty(values.get('name')) && 'Please enter a name';
  validation.dispositions =
    !isSaving &&
    !isFetching &&
    initialized &&
    (!values.get('dispositions') || values.get('dispositions').size === 0) &&
    'You must have one or more items in your Disposition List in order to save.';

  if (values.get('shared') === true && values.get('dispositions') && values.get('dispositions').size !== 0) {
    let unsharedDispositions = values.get('dispositions').filter(d => d.get('shared') !== true);
    if (unsharedDispositions.size !== 0) {
      validation.dispositions = "Shared disposition lists cannot contain dispositions that aren't also shared.";
    }
  }
  return validation;
};
