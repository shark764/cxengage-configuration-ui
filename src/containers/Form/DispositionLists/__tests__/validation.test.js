import { formValidation } from '../validation';
import { Map, List } from 'immutable';

describe('formValidation', () => {
  it('returns proper object when required fields are provided', () => {
    const values = new Map({
      name: 'mockName',
      dispositions: new List([{ dispositionId: 'mockId' }])
    });
    const props = {
      isSaving: false,
      isFetching: false,
      initialized: true
    };
    expect(formValidation(values, props)).toMatchSnapshot();
  });
  it('returns proper object (errors) when required fields are not provided', () => {
    const values = new Map({
      name: '',
      dispositions: new List()
    });
    const props = {
      isSaving: false,
      isFetching: false,
      initialized: true
    };
    expect(formValidation(values, props)).toMatchSnapshot();
  });
  it('returns proper object (errors) when wrong values are provided', () => {
    const values = new Map({
      name: undefined,
      dispositions: null
    });
    const props = {
      isSaving: false,
      isFetching: false,
      initialized: true
    };
    expect(formValidation(values, props)).toMatchSnapshot();
  });
  it('does not return disposition error when no dispositions are provided while saving', () => {
    const values = new Map({
      name: undefined,
      dispositions: null
    });
    const props = {
      isSaving: true,
      isFetching: false,
      initialized: true
    };
    expect(formValidation(values, props)).toMatchSnapshot();
  });
  it('does not return disposition error when no dispositions are provided while fetching', () => {
    const values = new Map({
      name: undefined,
      dispositions: null
    });
    const props = {
      isSaving: false,
      isFetching: true,
      initialized: false
    };
    expect(formValidation(values, props)).toMatchSnapshot();
  });
});
