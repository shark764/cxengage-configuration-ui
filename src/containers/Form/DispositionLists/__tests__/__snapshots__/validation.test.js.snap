// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formValidation returns proper object (errors) when required fields are not provided 1`] = `
Object {
  "dispositions": false,
  "name": "Please enter a name",
}
`;

exports[`formValidation returns proper object (errors) when wrong values are provided 1`] = `
Object {
  "dispositions": "You must have one or more items in your Disposition List in order to save.",
  "name": "Please enter a name",
}
`;

exports[`formValidation returns proper object when required fields are provided 1`] = `
Object {
  "dispositions": false,
  "name": false,
}
`;
